### ISSUE TITLE
invalid XML characters present in abstracts

### ISSUE THREAD

**Issue by alvacouch** [hydroshare/hydroshare#2737]
@martinseul @aphelionz @dtarb 
When running `check_bag` on all resources, the resource below crashed on bag creation, because there are invalid characters in the resource abstract when creating `resourcemetadata.xml`. Thus the bag for this resource cannot be created. The only way that this could happen is if the UI and/or REST API allowed the resource abstract to contain invalid characters. This was tested on cuahsi-dev-2.hydroshare.org using an image of the production system.

```
bag bags/1a7dc5d6b9fa4253bca442341eef500d.zip NOT FOUND
metadata_dirty is True
bag_modified is True
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/hydroshare/hs_core/management/commands/check_bag.py", line 194, in handle
    check_bag(r.short_id, options)
  File "/hydroshare/hs_core/management/commands/check_bag.py", line 48, in check_bag
    create_bag_files(resource)
  File "/hydroshare/hs_core/hydroshare/hs_bagit.py", line 92, in create_bag_files
    out.write(resource.get_metadata_xml())
  File "/hydroshare/hs_core/models.py", line 1949, in get_metadata_xml
    include_format_elements=include_format_elements)
  File "/hydroshare/hs_core/models.py", line 3935, in get_xml
    dcterms_abstract.text = self.description.abstract
  File "src/lxml/lxml.etree.pyx", line 1031, in lxml.etree._Element.text.__set__ (src/lxml/lxml.etree.c:51697)
  File "src/lxml/apihelpers.pxi", line 711, in lxml.etree._setNodeText (src/lxml/lxml.etree.c:23071)
  File "src/lxml/apihelpers.pxi", line 699, in lxml.etree._createTextNode (src/lxml/lxml.etree.c:22931)
  File "src/lxml/apihelpers.pxi", line 1439, in lxml.etree._utf8 (src/lxml/lxml.etree.c:30219)
ValueError: All strings must be XML compatible: Unicode or ASCII, no NULL bytes or control characters
```
I have modified `check_bag` to watch for this error and am running it again on all resources to check the extent of the bug.

**Comment by alvacouch** [comment]
Report from updated check_bag that doesn't crash: 
```
value error encountered for 1a7dc5d6b9fa4253bca442341eef500d: All strings must be XML compatible: Unicode or ASCII, no NULL bytes or control characters
value error encountered for 31f30d14c88748d986842d278d125a5c: All strings must be XML compatible: Unicode or ASCII, no NULL bytes or control characters.
```

**Comment by alvacouch** [comment]
See https://www.safaribooksonline.com/library/view/python-cookbook-2nd/0596007973/ch01s24.html
for one way to convert Unicode to XML in a lossless manner.

**Comment by alvacouch** [comment]
@dtarb @mseul @aphelionz @horsburgh 
I analyzed this situation in detail, and there were control characters in each resource's data, including ASCII code 1 (control-A) and ASCII code 11 (control-K). I have written a patch to remove these from the input to `resourcemetadata.xml`. 

But in looking at this, I inadvertently discovered that **all formatting is stripped out of the abstract when it is encoded.** For example, 1a7dc5d6b9fa4253bca442341eef500d has line breaks that don't make it into `resourcemetadata.xml`.  This makes things look bad, because the line breaks are not replaced by spaces; they are just missing. This leads to bad breaks like 
```
...Rapid Communications in Mass Spectrometry, 31(1), 75-84.Abstract:RATIONALE: The fast ...
```
instead of 
```
...Rapid Communications in Mass Spectrometry, 31(1), 75-84.

Abstract:

RATIONALE: The fast ...
```

**Is this intentional?  If not, what should the XML contain?**

At the very least, I think I need to substitute spaces for line breaks, e.g., 
```
...Rapid Communications in Mass Spectrometry, 31(1), 75-84. Abstract: RATIONALE: The fast ...
```

(The actual filtering is done by the xml tree functions. In order to avoid it, the formatting has to be pre-translated into its XML equivalents.)

**Comment by alvacouch** [comment]
@dtarb @aphelionz @martinseul @horsburgh  How is this for a compromise? Replace `\n` with paragraph symbol  `¶`. Replace control characters with spaces. Avoid repeating spaces or paragraph symbols. To wit:

```
… Rapid Communications in Mass Spectrometry, 31(1), 75-84. ¶ Abstract: ¶ RATIONALE: The fast ... ```

**Comment by dtarb** [comment]
@alvacouch I think we need a solution that is:
1. Simple
2. Consistent with general best practices in the field for working with XML files
3. Is lossless in that it would permit the abstract to be read back in from the XML identical to the way it is initially stored (not necessarily the same as the user input).
4. Gives the user flexibility to introduce formatting.  Users are always going to want to provide more formatting than is possible.  We should be as accommodating as is reasonable.  If it is possible to have some sort of rich text or markdown as is quite common in web content entry these days we should provide it, but only consistent with 1-3.
I realize that point 4 may lead us down paths that differ from 1-3, and that the definitions of 1 and 2 are imprecise.  

I tried this out a bit on dev-2.  See https://cuahsi-dev-2.hydroshare.org/resource/b6aa4bcafb90425eba889c3b4646731e/.  

Formatting with new lines (cut and paste from this comment) did not make it into the XML.
![image](https://user-images.githubusercontent.com/3456833/40112900-cd127684-58c4-11e8-8372-a389bf5d1c68.png).

I am pleased to see that special characters are accepted in the title and in the abstract.  Many of the considerations for abstract will apply to title and perhaps other fields too, though I think that there is no need for new lines in a title (and they appear not possible now - good). 


, though the landing page display is OK.

**Comment by alvacouch** [comment]
@dtarb So, you've found a case in which the formatting still gets stripped. Will investigate. My current solution only preserves newlines in otherwise bare text.

**Comment by horsburgh** [comment]
@dtarb - @alvacouch is right. We could potentially spend a lot of time here. It's an Abstract field - not an entire document. If we find that people need complex formatting to cram stuff into the Abstract field it is likely because we have failed somewhere else in letting them enter what they want - e.g., they want to put it in the Abstract because there isn't another place for it to go.

How much formatting do we really need to support?  Your number 1 isn't really consistent with numbers 3 and 4.

**Comment by alvacouch** [comment]
@dtarb @horsburgh You found is a bug in my bag generation. I regenerated the bag and it puts paragraph marks in where you might want.  So, I need to change the code for "new resourcemetadata.xml" versus "existing resourcemetadata.xml".  What I get now is: 

```
1. Simple ¶ 2. Consistent with general best practices in the field for working with XML files. ¶ compositions (δ2H and δ18O values) of soil and sediment ¶ 3. Is lossless in that it would permit the abstract to be read back in from the XML identical to the way it is initially stored (not necessarily the same as the user input). https://cuahsi-dev-2.hydroshare.org/resource/b6aa4bcafb90425eba889c3b4646731e/ ¶ 4. Gives the user flexibility to introduce formatting. Users are always going to want to provide more formatting than is possible. We should be as accommodating as is reasonable. If it is possible to have some sort of rich text or markdown as is quite common in web content entry these days we should provide it, but only consistent with 1-3. ¶ I realize that point 4 may lead us down paths that differ from 1-3, and that the definitions of 1 and 2 are imprecise
```

My guess is that to reconstruct what was there before, replacing carriage returns with paragraph marks meets your specifications and looks a lot better than what we have now. I would put whitespace before and after to make these prettier, but in spirit, we could reconstruct the original from the formatted copy.  

I note that 

* in the UI, `<p> .... </p>` are inserted wherever there is a newline.
* the resourcemetadata.xml creation code currently interprets the abstract as unicode text without html or xml markup.

**Comment by alvacouch** [comment]
@dtarb @horsburgh The bottom line is that there are only two options: 

1. Replace newlines with single spaces. 
2. Replace newlines with some other unicode character, e.g., ¶.

In existing code, newlines are simply ignored. This is unacceptable.  

Which will it be?

**Comment by alvacouch** [comment]
@dtarb @horsburgh It turns out that unicode has a special character ␤ that is specifically intended to indicate a newline in a stream of text. This is another option.

**Comment by alvacouch** [comment]
@dtarb @horsburgh  cuahsi-dev-2 is now working properly. Forgot to restart to flush .pyc usage.

**Comment by alvacouch** [comment]
@dtarb @horsburgh FYI: In the UI: isolated newlines are translated to <br> while sequences of newlines are translated to <p>. So, to be completely compatible with the UI, I would have to use both ␤ and ¶.

**Comment by dtarb** [comment]
@alvacouch please fix as you can right now using the best quick approach.

@horsburgh with respect to what people are trying to put in the abstract, if you just look at how people are trying to put information in it, I think it speaks to a need for greater flexibility.  I am all for keeping the abstract simple, and for fixing what we have now so that bag creation works, but over the longer term as part of advancing our resource data model I think we need to be considering text fields that users can add to describe their resources using richer formatting.  

I looked to see what the XML generated from the resource with formatting and special characters would be, and it is 
![image](https://user-images.githubusercontent.com/3456833/40175557-b6676940-5995-11e8-911c-a18954ae0c99.png)
If I cut and paste this into a new abstract I get
![image](https://user-images.githubusercontent.com/3456833/40175580-ce1bb302-5995-11e8-932a-ec19d2667abd.png)
Thus it is not clear how a user could recreate the formatted abstract from the encoding that has been embedded in the XML.  For now this may not be necessary, but longer term we need to think this through.  Some websites have view source options that allow tweaking of these sort of things.  We should be looking at what standard practices there are for these things.

**Comment by alvacouch** [comment]
@horsburgh @dtarb @aphelionz @martinseul Try http://cuahsi-dev-2.hydroshare.org now. 

    - Multiple newline sequences are replaced with ¶ (Unicode paragraph symbol) 
    - Isolated newlines (e.g., a line break) are replaced with ⏎ (Unicode symbol for carriage return)
    - These are padded with spaces. 
    - Either one or the other appears. There are no combinations. 

This was in the spirit of the call yesterday.

**Comment by horsburgh** [comment]
@dtarb - I'm not questioning the need for enhanced information, I'm saying that the Abstract isn't necessarily the place for it. Many people may be cramming stuff in the Abstract that could/should go in other metadata elements because they either don't know about what is available to them or they don't want to take the time to fill things out. Some best practices here may alleviate the need for complex functionality around what should be a simple text entry field.

Certainly a review of what has been done by users will clue us in on how people are using the Abstract field and where improvements can be made.

**Comment by dtarb** [comment]
@alvacouch what is the status of this.  If you are busy with downloads, should we assign this to someone else.

**Comment by alvacouch** [comment]
@dtarb The proposed fix is done and part of the "Large Files" PR. The algorithm is simple: replace two or more newlines with a paragraph; one with a newline character in UML. This matches the way abstracts are shown in the UI: two or more means <p>, while one in isolation means <br>.

**Comment by dtarb** [comment]
@alvacouch @aphelionz not sure if this was supposed to be addressed in 1.16 deployed today, but https://www.hydroshare.org/resource/1fedd80cd20245409a0ea02cc2a1ecf0/ is broken for downloading a bag and for copying.  When attempting to copy the error is 

![image](https://user-images.githubusercontent.com/3456833/47385527-87cd1300-d6c7-11e8-929d-39906e53547a.png)

**Comment by martinseul** [comment]
@hyi @sblack-usu @alvacouch i checked and this is still an issue in production

**Comment by martinseul** [comment]
close, fix is implemented with #3586
